### PR TITLE
Update Uncontrolled NOx and SO2 factors

### PR DESCRIPTION
**Pull request by grgmiller**
### Purpose
This PR was started to add NOx and SO2 factors for boiler configurations in the historical data that were not previously specified. This is part of the effort to improve NOx and SO2 data quality for new historical years from 2005-2018.

**Programmatic creation of the factor tables**
Instead of continuing to manually update these in the table, we add a script to programmatically create the nox and so2 factor reference tables to address https://github.com/singularity-energy/open-grid-emissions/issues/276. This is meant to reduce the risk of typos, ensure consistency in the methodology, and also add transparency to how these factors are created.

**Fluidized Bed Boilers**
As noted by the EIA's Electric Power Annual [technical notes](https://www.eia.gov/electricity/annual/pdf/tech_notes.pdf), while the uncontrolled emissions factors published by EIA need to be adjusted to account for each plant's operational pollution control equipment, the exception is for fluidized bed boilers: 
>A special case for removal of SO2 is the fluidized bed boiler, in which the sulfur removal process is integral with the operation of the boiler. The SO2 emission factors shown in Table A.1. for fluidized bed boilers already account for 90 percent removal of SO2 since, in effect, the plant has no uncontrolled emissions of this pollutant.

Previously, we had been applying pollution control efficiency values to adjust so2 emissions from these boilers, which would underestimate emissions. We now do not adjust so2 emissions for this type of boiler since the calculated values already represent controlled emissions. It looks like this may have only affected a small number of plants, since many plants did not report control efficiencies for fluidized bed technologies in EIA-923.

**Update default assumed value**
This PR also includes an important methodological update on the assumptions used for assigning NOx and SO2 factors for boilers with missing boiler configuration data. When boiler configuration data is unknown, we had previously been using the lowest fuel-specific NOx or SO2 factor of any boiler firing type, with the reasoning that this is the minimum amount of emissions that could be expected. This was also the approach that eGRID had used in previous years (looking at more recent versions of eGRID, it looks like they may have abandoned this approach). 

However, there are several drawbacks of this approach. For SO2, since the fluidized bed factors are already controlled, this approach would mean generally assuming that the generator is a fluidized bed boiler, which would be applying a special case as the default backstop. In the case of NOx, this could lead to the assumptions of different types of boilers being the standard default. The EIA's default assumption, in the case that boiler information is unavailable, is to assume that the boiler has a dry-bottom firing method falling into the "all other category." We adopt this assumption in this PR, which generally have the effect of increasing NOx and SO2 emissions in these cases.
https://github.com/singularity-energy/open-grid-emissions/issues/150

### What the code is doing
How is the purpose executed?

### Testing
In the notebook, I test for:
- duplicate factors
- How much the current factors differ from the factors already in `reference_tables`
- That all historical boiler configurations are covered.

Ran pipeline for 2017-2018

### Where to look
- src/oge/create_nox_so2_factors.py

### Usage Example/Visuals
For NOx, there are 80 factors that are different from previous.
image

For SO2, there are 288 factors that are different from the previous factors.
image

Inspecting these differences reveals that they are based on the updated default assumption, and how we treat combustion turbines with a specified boiler firing type. Where previously we had been assigning the combustion turbine factor, we now use the factor associated with the boiler firing type, since these may be associated with HRSG configurations (see: https://www.eia.gov/todayinenergy/detail.php?id=52778)

There are a few configurations for which NOx factors are still missing, but these seem to be anomalies (eg a hydro plant burning oil). This may just be a figment of the test logic I set up and these anomalous configurations may not appear in the real data.
image

### Review estimate
20-30 min

### Future work
Updating LFG factors: https://github.com/singularity-energy/open-grid-emissions/issues/218
Default factor assumptions: https://github.com/singularity-energy/open-grid-emissions/issues/69

### Checklist
- [x] Update the documentation to reflect changes made in this PR
- [x] Format all updated python files using `black`
- [x] Clear outputs from all notebooks modified
- [x] Add docstrings and type hints to any new functions created

### CODE REVIEW

File: src/oge/load_data.py [hunk 1]
```diff
@@ -1224,6 +1202,9 @@ def load_emissions_controls_eia923(year: int) -> pd.DataFrame:
         emissions_controls_eia923["report_date"] = emissions_controls_eia923[
             "report_date"
         ].astype("datetime64[s]")
+        emissions_controls_eia923["equipment_tech_description"] = (
+            emissions_controls_eia923["equipment_tech_description"].str.lower()
+        )
     else:
         logger.warning(
             "Emissions control data prior to 2012 has not been integrated into the data pipeline."
```
Comment (rouille) [inline src/oge/load_data.py:1210]: Is that 2012 or 2008?

File: src/oge/create_nox_so2_factors.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+    There are several filling steps:
+    1. Missing energy_source_codes: The EIA tables do not provide NOx or SO2 factors for
+    MSB, MSN, or SC fuel types. For MSB and MSN (the biogenic and nonbiogenic
+    of MSW), we set the factors equal to the specified MSW factor. For SC (Coal Synfuel)
+    We set the factor equal to the refined coal (RC) factor, which itself is assumed to
+    be equal to the bituminous coal (BIT) factor.
+    2. Factors for units missing boiler_firing_type (BFT) and wet_dry_bottom (WDB) data:
+    According to the technical appendix for the EIA's Electric Power Annual (see:
+    https://www.eia.gov/electricity/annual/pdf/tech_notes.pdf), whenever a unit is
+    missing BFT and WDB data (in the case of NOx), they apply they assume it is a dry-
```
Comment (rouille) [inline src/oge/create_nox_so2_factors.py:313]: they apply they assume -> they assume

File: src/oge/create_nox_so2_factors.py [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+            "data_source",
+            "unit",
+            "cyclone_firing",
+            "fluidized_bed_firing",
+            "stoker",
+            "tangential_firing_dry",
+            "tangential_firing_wet",
+            "other_dry",
+            "other_wet",
+            "GT",
```
Comment (rouille) [inline src/oge/create_nox_so2_factors.py:51]: Should it be `CT` for combustion turbine? Just want to make sure it is not a typo

File: src/oge/create_nox_so2_factors.py [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+
+    factor_columns = [
+        "cyclone_firing",
+        "fluidized_bed_firing",
+        "stoker",
+        "tangential_firing_dry",
+        "tangential_firing_wet",
+        "other_dry",
+        "other_wet",
+        "GT",
```
Comment (rouille) [inline src/oge/create_nox_so2_factors.py:65]: Same comment here `GT` -> `CT`

File: src/oge/create_nox_so2_factors.py (hunk 3 above)
Comment (grgmiller) [reply src/oge/create_nox_so2_factors.py:51]: If you look at the prime mover metadata (https://catalystcoop-pudl.readthedocs.io/en/latest/data_dictionaries/codes_and_labels.html#core-eia-codes-prime-movers), GT is the code for a general combustion turbine, while CT is specifically a combustion turbine that is part of a combined cycle unit. 
We end up applying this EF to the CT prime mover later in the pipeline.

File: src/oge/create_nox_so2_factors.py (hunk 4 above)
Comment (grgmiller) [reply src/oge/create_nox_so2_factors.py:65]: See above comment

File: src/oge/create_nox_so2_factors.py (hunk 2 above)
Comment (grgmiller) [reply src/oge/create_nox_so2_factors.py:313]: Fixed

File: src/oge/load_data.py (hunk 1 above)
Comment (grgmiller) [reply src/oge/load_data.py:1210]: PUDL has integrated data from 2008 onward, but their version of the table does not currently include columns for hours_in_service and ozone_season_nox_emission_rate. See: https://github.com/catalyst-cooperative/pudl/issues/3689. 

We are still manually loading the EIA-923 table from a raw download, and we have only integrated back to 2012 so far.

File: src/oge/create_nox_so2_factors.py [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+    # apply the combustion turbine factor to all combustion-turbine-type prime movers,
+    # including those at combined cycle plants, and CAES plants
+    for pm in ["CA", "CC", "CS", "CT", "CE"]:
+        df[pm] = df["GT"]
+
+    # where we have a prime mover with no BFT or WDB information, we assume a dry-bottom
+    # boiler in the "all other" BFT category. This is consistent with the assumption made
+    # by the EIA. see: https://www.eia.gov/electricity/annual/pdf/tech_notes.pdf
+    # for pm in ["ST","OT"]:
+    #    df[pm] = df[default_value]
```
Comment (rouille) [inline src/oge/create_nox_so2_factors.py:374]: Why is that commented? Don't we want to assign factors for these PMs?

**Review by rouille — APPROVED** [review]
Thanks

**Pull request merged**